File: examples/rust/src/bin/10_iterators.rs
//! Example 10 — Iterator APIs
//!
//! Tensogram provides three iterator types for ergonomic traversal:
//!
//!   messages(&buf)   — zero-copy iterator over messages in a byte buffer
//!   objects(&msg, _) — lazy iterator decoding each object on demand
//!   file.iter()      — lazy seek-based iteration over file messages
//!
//! All three implement Iterator and ExactSizeIterator.

use std::collections::BTreeMap;

use ciborium::Value;
use tensogram_core::{
    decode, messages, objects, objects_metadata, ByteOrder, DataObjectDescriptor, DecodeOptions,
    Dtype, EncodeOptions, GlobalMetadata, TensogramFile,
};

fn make_message(param: &str, fill: u8) -> (GlobalMetadata, DataObjectDescriptor, Vec<u8>) {
    let mars = Value::Map(vec![(
        Value::Text("param".into()),
        Value::Text(param.into()),
    )]);
    let mut entry = BTreeMap::new();
    entry.insert("mars".to_string(), mars);

    let global_meta = GlobalMetadata {
        version: 2,
        base: vec![entry],
        ..Default::default()
    };

    let desc = DataObjectDescriptor {
        obj_type: "ntensor".to_string(),
        ndim: 2,
        shape: vec![4u64, 4],
        strides: vec![4u64, 1],
        dtype: Dtype::Float32,
        byte_order: ByteOrder::Little,
        encoding: "none".to_string(),
        filter: "none".to_string(),
        compression: "none".to_string(),
        params: BTreeMap::new(),
        hash: None,
    };

    let data = vec![fill; 4 * 4 * 4]; // 4×4 float32
    (global_meta, desc, data)
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    // ── 1. Buffer message iteration ─────────────────────────────────────────
    //
    // Encode 3 messages, concatenate into a single buffer, then iterate.
    let params = ["2t", "10u", "msl"];
    let encoded: Vec<Vec<u8>> = params
        .iter()
        .enumerate()
        .map(|(i, &p)| {
            let (meta, desc, data) = make_message(p, i as u8);
            tensogram_core::encode(&meta, &[(&desc, &data)], &EncodeOptions::default()).unwrap()
        })
        .collect();

    let mut buf = Vec::new();
    for msg in &encoded {
        buf.extend_from_slice(msg);
    }

    println!("=== Buffer message iterator ===");
    let iter = messages(&buf);
    println!("  {} messages found (ExactSizeIterator)", iter.len());

    for (i, msg_slice) in messages(&buf).enumerate() {
        let (meta, _) = decode(msg_slice, &DecodeOptions::default())?;
        let param = meta
            .base
            .first()
            .and_then(|e| e.get("mars"))
            .map(|m| get_text(m, "param"))
            .unwrap_or("?");
        println!("  [{i}] param={param}  bytes={}", msg_slice.len());
    }

    // ── 2. Object iteration ─────────────────────────────────────────────────
    //
    // Iterate over the decoded objects inside a single message.
    println!("\n=== Object iterator ===");
    let first_msg = messages(&buf).next().unwrap();
    for result in objects(first_msg, DecodeOptions::default())? {
        let (desc, data) = result?;
        println!(
            "  shape={:?}  dtype={}  data={} bytes",
            desc.shape,
            desc.dtype,
            data.len()
        );
    }

    // ── 3. Metadata-only iteration ──────────────────────────────────────────
    //
    // Get descriptors without decoding any payload data.
    println!("\n=== Metadata-only ===");
    for desc in objects_metadata(first_msg)? {
        println!(
            "  ndim={}  shape={:?}  dtype={}",
            desc.ndim, desc.shape, desc.dtype
        );
    }

    // ── 4. File-based iteration ─────────────────────────────────────────────
    //
    // Write messages to a file, then iterate lazily with seek-based I/O.
    let dir = tempfile::tempdir()?;
    let path = dir.path().join("iter_demo.tgm");

    {
        let mut file = TensogramFile::create(&path)?;
        for (i, &p) in params.iter().enumerate() {
            let (meta, desc, data) = make_message(p, i as u8);
            file.append(&meta, &[(&desc, &data)], &EncodeOptions::default())?;
        }
    }

    println!("\n=== File iterator ===");
    let file = TensogramFile::open(&path)?;
    let file_iter = file.iter()?;
    println!("  {} messages in file (ExactSizeIterator)", file_iter.len());

    for (i, raw) in file_iter.enumerate() {
        let raw = raw?;
        let (meta, _) = decode(&raw, &DecodeOptions::default())?;
        let param = meta
            .base
            .first()
            .and_then(|e| e.get("mars"))
            .map(|m| get_text(m, "param"))
            .unwrap_or("?");
        println!("  [{i}] param={param}  bytes={}", raw.len());

        // Nested: iterate objects within this message
        for result in objects(&raw, DecodeOptions::default())? {
            let (desc, data) = result?;
            println!(
                "       shape={:?}  dtype={}  {} bytes",
                desc.shape,
                desc.dtype,
                data.len()
            );
        }
    }

    println!("\nIterator example complete.");
    Ok(())
}

fn get_text<'a>(map: &'a Value, key: &str) -> &'a str {
    if let Value::Map(entries) = map {
        for (k, v) in entries {
            if matches!(k, Value::Text(s) if s == key) {
                if let Value::Text(t) = v {
                    return t;
                }
            }
        }
    }
    "?"
}

File: examples/typescript/10_iterators.ts
// (C) Copyright 2026- ECMWF and individual contributors.
//
// This software is licensed under the terms of the Apache Licence Version 2.0
// which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
// In applying this licence, ECMWF does not waive the privileges and immunities
// granted to it by virtue of its status as an intergovernmental organisation nor
// does it submit to any jurisdiction.

/**
 * Example 10 — Iterator APIs (TypeScript)
 *
 * Three ways to walk a multi-message `.tgm` without reading it all into
 * memory at once:
 *
 *   1. `for await (const msg of file)` — file-level async iteration.
 *   2. `scan(buf)` — buffer-level message-position scan, tolerant of
 *      corrupt regions between messages.
 *   3. `file.message(i)` / `file.messageMetadata(i)` — random-access
 *      by index (O(1) via the binary header).
 */

import {
  decode,
  encode,
  init,
  scan,
  TensogramFile,
  type DataObjectDescriptor,
  type GlobalMetadata,
} from '@ecmwf.int/tensogram';
import { mkdtempSync, rmSync, writeFileSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { join } from 'node:path';

function describe(shape: number[]): DataObjectDescriptor {
  const strides = new Array<number>(shape.length).fill(1);
  for (let i = shape.length - 2; i >= 0; i--) strides[i] = strides[i + 1] * shape[i + 1];
  return {
    type: 'ntensor',
    ndim: shape.length,
    shape,
    strides,
    dtype: 'float32',
    byte_order: 'little',
    encoding: 'none',
    filter: 'none',
    compression: 'none',
  };
}

function makeMessage(param: string, step: number): Uint8Array {
  const metadata: GlobalMetadata = {
    version: 3,
    base: [{ mars: { param, step } }],
  };
  const data = new Float32Array(10).fill(step + 0.5);
  return encode(metadata, [{ descriptor: describe([10]), data }]);
}

async function main(): Promise<void> {
  await init();

  const tmp = mkdtempSync(join(tmpdir(), 'tgm-iter-'));
  const path = join(tmp, 'forecast.tgm');
  try {
    // ── 1. Write a file of four messages ────────────────────────────────
    const messages = [
      makeMessage('2t', 0),
      makeMessage('10u', 0),
      makeMessage('2t', 6),
      makeMessage('10u', 6),
    ];
    const concatenated = new Uint8Array(messages.reduce((n, m) => n + m.byteLength, 0));
    let offset = 0;
    for (const m of messages) {
      concatenated.set(m, offset);
      offset += m.byteLength;
    }
    writeFileSync(path, concatenated);
    console.log(`Wrote ${concatenated.byteLength} bytes, ${messages.length} messages`);

    // ── 2. Async iteration over the file ────────────────────────────────
    //
    // `TensogramFile` implements `AsyncIterable<DecodedMessage>`, so
    // `for await` is the idiomatic walk.  Each yielded handle owns
    // WASM-side memory and must be `.close()`'d.
    console.log('\n1. for await (const msg of file):');
    const file = await TensogramFile.open(path);
    try {
      let i = 0;
      for await (const msg of file) {
        try {
          const param = (msg.metadata.base?.[0] as { mars?: { param: string } } | undefined)
            ?.mars?.param ?? '?';
          console.log(`   [${i++}] param=${param}  objects=${msg.objects.length}`);
        } finally {
          msg.close();
        }
      }
    } finally {
      file.close();
    }

    // ── 3. Buffer-level scan ─────────────────────────────────────────────
    //
    // `scan` finds message boundaries in an arbitrary byte buffer — you
    // can scan bytes that didn't come from a file (HTTP body, socket,
    // pipe, clipboard, ...).  A corrupt stretch between messages is
    // skipped silently; `scan` only reports the positions it could
    // verify end-to-end.
    console.log('\n2. scan(buf) over an in-memory buffer:');
    const positions = scan(concatenated);
    for (let j = 0; j < positions.length; j++) {
      const { offset: o, length: len } = positions[j];
      const slice = concatenated.subarray(o, o + len);
      const msg = decode(slice);
      try {
        const param = (msg.metadata.base?.[0] as { mars?: { param: string } } | undefined)
          ?.mars?.param ?? '?';
        console.log(`   [${j}] offset=${o}  length=${len}  param=${param}`);
      } finally {
        msg.close();
      }
    }

    // ── 4. Random access by index ────────────────────────────────────────
    //
    // `file.message(i)` seeks directly to message i using the offset
    // map built during the initial scan.  Combine with
    // `file.messageMetadata(i)` for metadata-only inspection.
    console.log('\n3. Random access — file.message(i):');
    const file2 = await TensogramFile.open(path);
    try {
      const first = await file2.messageMetadata(0);
      const last = await file2.messageMetadata(file2.messageCount - 1);
      const paramOf = (m: GlobalMetadata) =>
        ((m.base?.[0] as { mars?: { param: string } } | undefined)?.mars?.param ?? '?');
      console.log(`   first: param=${paramOf(first)}`);
      console.log(`   last : param=${paramOf(last)}`);

      const mid = await file2.message(1);
      try {
        const shape = mid.objects[0].descriptor.shape;
        console.log(`   file.message(1).objects[0].shape = [${shape.join(', ')}]`);
      } finally {
        mid.close();
      }
    } finally {
      file2.close();
    }
  } finally {
    rmSync(tmp, { recursive: true });
  }

  console.log('\nIterator APIs OK.');
}

main().catch((err: unknown) => {
  console.error(err);
  process.exit(1);
});
